### PR TITLE
Remove the default block margins from themes with theme.json file

### PR DESCRIPTION
**Pull request by youknowriad**
Tries to implement the idea expressed by @jasmussen here https://github.com/WordPress/gutenberg/issues/27315#issuecomment-810055517

This PR removes the default block margins applied in the editor from themes with experimenta-theme.json file..
This ensures that these themes are more WSIWYG

### CODE REVIEW

File: packages/edit-post/src/components/layout/style.scss [hunk 1]
```diff
@@ -124,6 +124,11 @@
 		margin-left: auto;
 		margin-right: auto;
 
+		// Provide every block with a default base margin. This margin provides a consistent spacing
+		// between blocks in the editor.
+		margin-top: $default-block-margin;
+		margin-bottom: $default-block-margin;
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/style.scss]: This is now has a bit more specificity than the previous block margins for classic themes. This is probably the most impactful thing but I don't think we have another way to apply these to only the classic themes without the added specificity. (There might be another way by adding an inline style in php but I'd rather not do it unless necessary :P)

**Comment by jasmussen** [comment]
Thanks so much for trying this. As noted, those default margin values, but top and bottom, left and right, have caused headaches for a long time.

This PR is a step in the right direction for block themes, but the increased specificity of the .wp-block selector as you note, is causing some headaches even just in this branch, as they now override theme specificity in all the themes I tried:

Screenshot 2021-03-30 at 12 36 35

If we were able to output the old block styles in a separate stylesheet that was not loaded for block themes, essentially a back-compat style, that seems like it would be able to keep the legacy specificity intact, while not burdening block themes. Is this what you'd prefer to avoid (based on the second comment)?

One alternative I can think of is to invert the styles, to something like this:

```
.wp-block {
	margin-left: auto;
	margin-right: auto;
	margin-top: $default-block-margin;
	margin-bottom: $default-block-margin;
}

.supports-layout .wp-block {
	margin-left: revert;
	margin-right: revert;
	margin-top: revert;
	margin-bottom: revert;
}
```

But that puts the extra specificity burden on block themes.

A 3rd option is to load a separate additional stylesheet just for block themes, one that includes only these styles:

```
.wp-block {
	margin-left: revert;
	margin-right: revert;
	margin-top: revert;
	margin-bottom: revert;
}
```

The only option of those three that I like, is to output the legacy styles in a separate stylesheet, loaded conditionally. What's in trunk currently is already a bit of a headache (see again https://github.com/WordPress/gutenberg/pull/30337).

**Comment by youknowriad** [comment]
Thanks for the feedback @jasmussen I implemented your first suggestion, let me know what you think.

**Comment by carolinan** [comment]
This separate file is my preferred solution as well.
✅  `wp-editor-classic-layout-styles-css`, classic.css,  loads when I activate a non FSE theme.


Twenty Twenty One is broken.

Before, with current Gutenberg trunk only:
blocks are aligned correctly in the editor

With PR applied:
left aligned blocks are not aligned correctly in the editor

The sample content is the Block: Cover post from the theme unit test.

-----





✅ Random selection of blocks in an FSE theme, with no top and bottom margins:
blocks have no top or bottom margin in the site editor

**Comment by carolinan** [comment]
Seedlet (.org version) with PR applied, same content as above:
The blocks are not aligned correctly

Twenty Nineteen with PR applied, same content as above:
The blocks are almost aligned correctly, spacing is missing between cover block and paragraph, making text difficult to read.

**Comment by jasmussen** [comment]
Thanks so much Riad, this is working really well for me, and the split seems like it will provide us with a way to support both paths forward in as best a possible way. I think this'll work well, though I'd encourage additional thoughts. @mcsf?

I'm seeing the same issue as Carolina, even with the demo content:

Screenshot 2021-04-05 at 09 29 25

The rules are output correctly it appears:

Screenshot 2021-04-05 at 09 29 49

The new lower specificity (which, amazing, by the way), now gets overridden by this:

Screenshot 2021-04-05 at 09 30 05

Perhaps we can lower the specificity on that reset as well?

**Comment by youknowriad** [comment]
> Perhaps we can lower the specificity on that reset as well?

That's not possible, these are there to override wp-admin default styles.

**Comment by youknowriad** [comment]
The bug should be fixed now, I restored the specificity that was applied to these rules before #29335 and in trunk as well.

**Comment by jasmussen** [comment]
This is super duper solid. Here's TwentyTwentyOne in this branch:

Screenshot 2021-04-05 at 12 15 43

Buttons are good again:

Screenshot 2021-04-05 at 12 16 17

Here's trunk:

Screenshot 2021-04-05 at 12 17 12

**Comment by jasmussen** [comment]
Big fan of this approach. It could use a good bit of testing, but 👍 👍 from here. Thank you Riad!

**Review by ntsekouras — APPROVED** [review]
LGTM! 👍 - I tested with various themes and in general this feels so much better! Thanks Riad!

File: lib/client-assets.php [hunk 2]
```diff
@@ -360,6 +360,11 @@ function gutenberg_register_packages_styles( $styles ) {
 		'wp-reusable-blocks',
 	);
 
+	// Only load the default layout and margin styles for themes without theme.json file.
+	if ( ! WP_Theme_JSON_Resolver::theme_has_support() ) {
+		$wp_edit_blocks_dependencies[] = 'wp-editor-classic-layout-styles';
```
Comment (gziolo) [inline lib/client-assets.php:365]: This is such a nice way to ensure backward compatibility. ❤️

**Comment by youknowriad** [comment]
Looks like this PR fixes some issues in 10.3.1. If we do a 10.3.2 we could consider including this one.

**Comment by gziolo** [comment]
@WunderBart, it looks like we have 2 candidates for 10.3.2, including your https://github.com/WordPress/gutenberg/pull/30540.

**Pull request merged**